### DISCUSSION TITLE
android ，Is there a tutorial on how to call cameras and albums.

### DISCUSSION THREAD

**Discussion by jlij** [beeware/briefcase#1278] [Q&A] [answered]
I want to call the camera by clicking on the DetailedList，And assign the element information to the photo name。

**Comment by freakboy3742** [comment]
Unfortunately, we don't have a tutorial for this. We don't currently have a cross-platform API for these features; to implement them, you need to write code against the native platform APIs for camera access, etc. You'll need to combine the [documentation for Chaquopy](https://chaquo.com/chaquopy/doc/current/python.html) layer with Android's own platform documentation for camera and album access.

**Reply by jlij** [reply]
Will your team consider joining this API in the future? I am a beginner in Python, and I don't understand Java. I can only wait and see.

**Reply by freakboy3742** [reply]
Yes - adding a cross-platform API for camera and album access is on our todo list. However, I wouldn't expect to see any progress on this until late this year.

**Reply by jlij** [reply]
It's okay, I'll wait. Thank you very much for your patient answer.

**Reply by jlij** [reply]
Hello teacher, has there been any progress in the camera call function?

**Reply by freakboy3742** [reply]
Yes - a Camera API was added added in Toga 0.4.2, with support for iOS, Android and macOS. Example usage can be found in the [hardware example project](https://github.com/beeware/toga/tree/main/examples/hardware).

**Reply by jlij** [reply]
Prompt after startup:The Camera API is not implemented on this platform,
Click on Take photo:NotImplementedError: Toga's Winforms backend doesn't implement Camera.
How can I solve this problem?

**Reply by freakboy3742** [reply]
The error message isn't a cryptic puzzle. It's telling you the exact problem. The Camera API hasn't been implemented for Windows. 

The [platform support chart for the Camera API](https://toga.readthedocs.io/en/stable/reference/api/hardware/camera.html) also says this, as does my reply telling you that the Camera API is available:

> Yes - a Camera API was added added in Toga 0.4.2, with support for iOS, Android and macOS.

If you want to use the camera on Windows, you'll need to either implement the Toga API for Windows, or find a third-party library that can access the camera.

**Reply by jlij** [reply]
Do I need to define it in pyproject.toml? Is it okay to write it like this? Permission. camera="android. permission. CAMERA"

**Reply by freakboy3742** [reply]
This isn't a case of you doing something wrong. It doesn't require a 1 line configuration change. The feature *does not exist*. Toga does not contain an implementation of the Camera API on Windows or Linux. At present, you can only use the Camera API on macOS, iOS and Android. The "fix" is for someone to implement the entire Camera API for Winforms and GTK.

**Reply by jlij** [reply]
How to solve problems encountered in the Android environment?

--------- beginning of crash
E/AndroidRuntime: FATAL EXCEPTION: main
E/AndroidRuntime: Process: com.example.helloworld, PID: 29938
E/AndroidRuntime: java.lang.RuntimeException: Unable to start activity ComponentInfo{com.example.helloworld/org.beeware.android.MainActivity}: com.chaquo.python.PyException: AttributeError: 'ExampleHardwareApp' object has no attribute 'camera'
E/AndroidRuntime:       at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3303)
E/AndroidRuntime:       at android.app.ActivityThread.handleLaunchActivity(ActivityThread.java:3411)
E/AndroidRuntime:       at android.app.ActivityThread.-wrap12(Unknown Source:0)
E/AndroidRuntime:       at android.app.ActivityThread$H.handleMessage(ActivityThread.java:1994)
E/AndroidRuntime:       at android.os.Handler.dispatchMessage(Handler.java:108)
E/AndroidRuntime:       at android.os.Looper.loop(Looper.java:166)
E/AndroidRuntime:       at android.app.ActivityThread.main(ActivityThread.java:7529)
E/AndroidRuntime:       at java.lang.reflect.Method.invoke(Native Method)
E/AndroidRuntime:       at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:245)
E/AndroidRuntime:       at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:921)
E/AndroidRuntime: Caused by: com.chaquo.python.PyException: AttributeError: 'ExampleHardwareApp' object has no attribute 'camera'
E/AndroidRuntime:       at <python>.helloworld.app.startup(app.py:11)
E/AndroidRuntime:       at <python>.toga_android.app.create(app.py:175)
E/AndroidRuntime:       at <python>.toga_android.app.main_loop(app.py:186)
E/AndroidRuntime:       at <python>.toga.app.main_loop(app.py:517)
E/AndroidRuntime:       at <python>.__main__.<module>(__main__.py:4)
E/AndroidRuntime:       at <python>.runpy._run_code(<frozen runpy>:88)
E/AndroidRuntime:       at <python>.runpy._run_module_code(<frozen runpy>:98)
E/AndroidRuntime:       at <python>.runpy.run_module(<frozen runpy>:226)
E/AndroidRuntime:       at <python>.chaquopy_java.call(chaquopy_java.pyx:354)
E/AndroidRuntime:       at <python>.chaquopy_java.Java_com_chaquo_python_PyObject_callAttrThrowsNative(chaquopy_java.pyx:326)
E/AndroidRuntime:       at com.chaquo.python.PyObject.callAttrThrowsNative(Native Method)
E/AndroidRuntime:       at com.chaquo.python.PyObject.callAttrThrows(PyObject.java:232)
E/AndroidRuntime:       at com.chaquo.python.PyObject.callAttr(PyObject.java:221)
E/AndroidRuntime:       at org.beeware.android.MainActivity.onCreate(MainActivity.java:85)
E/AndroidRuntime:       at android.app.Activity.performCreate(Activity.java:7383)
E/AndroidRuntime:       at android.app.Instrumentation.callActivityOnCreate(Instrumentation.java:1218)
E/AndroidRuntime:       at android.app.ActivityThread.performLaunchActivity(ActivityThread.java:3256)
E/AndroidRuntime:       ... 9 more
I/Process : Sending signal. PID: 29938 SIG: 9

**Comment by freakboy3742** [comment]
That stack trace indicates you're not using Toga 0.4.2. You've either not updated your `pyproject.toml`, or you haven't run `briefcase update android -r` to update dependencies.

**Reply by jlij** [reply]
You can enter the interface normally after configuring Toga=0.4.2, but there is an error when clicking the button to call the camera.

V/AudioManager: playSoundEffect   effectType: 0
W/python.stderr: Error in async handler: Couldn't find meta-data for provider with authority org.beeware.examples.hardware.fileprovider
W/python.stderr:        at androidx.core.content.FileProvider.getFileProviderPathsMetaData(FileProvider.java:664)
W/python.stderr:        at androidx.core.content.FileProvider.parsePathStrategy(FileProvider.java:695)
W/python.stderr:        at androidx.core.content.FileProvider.getPathStrategy(FileProvider.java:645)
W/python.stderr:        at androidx.core.content.FileProvider.getUriForFile(FileProvider.java:449)
W/python.stderr:        at com.chaquo.python.PyObject.callAttrThrowsNative(Native Method)
W/python.stderr:        at com.chaquo.python.PyObject.callAttrThrows(PyObject.java:232)
W/python.stderr:        at com.chaquo.python.PyInvocationHandler.invoke(PyInvocationHandler.java:31)
W/python.stderr:        at java.lang.reflect.Proxy.invoke(Proxy.java:913)
W/python.stderr:        at $Proxy4.run(Unknown Source)
W/python.stderr:        at android.os.Handler.handleCallback(Handler.java:808)
W/python.stderr:        at android.os.Handler.dispatchMessage(Handler.java:101)
W/python.stderr:        at android.os.Looper.loop(Looper.java:166)
W/python.stderr:        at android.app.ActivityThread.main(ActivityThread.java:7529)
W/python.stderr:        at java.lang.reflect.Method.invoke(Native Method)
W/python.stderr:        at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:245)
W/python.stderr:        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:921)
W/python.stderr: Traceback (most recent call last):
W/python.stderr:   File "/data/data/com.example.helloworld/files/chaquopy/AssetFinder/requirements/toga/handlers.py", line 38, in handler_with_cleanup
W/python.stderr:   File "/data/data/com.example.helloworld/files/chaquopy/AssetFinder/app/helloworld/app.py", line 54, in take_photo
W/python.stderr:   File "/data/data/com.example.helloworld/files/chaquopy/AssetFinder/requirements/toga/hardware/camera.py", line 122, in take_photo
W/python.stderr:   File "/data/data/com.example.helloworld/files/chaquopy/AssetFinder/requirements/toga_android/hardware/camera.py", line 97, in take_photo
W/python.stderr:   File "class.pxi", line 972, in java.chaquopy.JavaMultipleMethod.__get__.lambda3
W/python.stderr:   File "class.pxi", line 1010, in java.chaquopy.JavaMultipleMethod.__call__
W/python.stderr:   File "class.pxi", line 791, in java.chaquopy.JavaMethod.__call__
W/python.stderr:   File "class.pxi", line 954, in java.chaquopy.JavaMethod.call_static_method
W/python.stderr:   File "env.pxi", line 262, in java.chaquopy.CQPEnv.CallStaticObjectMethodA
W/python.stderr:   File "env.pxi", line 546, in java.chaquopy.CQPEnv.check_exception
W/python.stderr: java.lang.IllegalArgumentException: Couldn't find meta-data for provider with authority org.beeware.examples.hardware.fileprovider
W/python.stderr:        at androidx.core.content.FileProvider.getFileProviderPathsMetaData(FileProvider.java:664)
W/python.stderr:        at androidx.core.content.FileProvider.parsePathStrategy(FileProvider.java:695)
W/python.stderr:        at androidx.core.content.FileProvider.getPathStrategy(FileProvider.java:645)
W/python.stderr:        at androidx.core.content.FileProvider.getUriForFile(FileProvider.java:449)
W/python.stderr:        at com.chaquo.python.PyObject.callAttrThrowsNative(Native Method)
W/python.stderr:        at com.chaquo.python.PyObject.callAttrThrows(PyObject.java:232)
W/python.stderr:        at com.chaquo.python.PyInvocationHandler.invoke(PyInvocationHandler.java:31)
W/python.stderr:        at java.lang.reflect.Proxy.invoke(Proxy.java:913)
W/python.stderr:        at $Proxy4.run(Unknown Source)
W/python.stderr:        at android.os.Handler.handleCallback(Handler.java:808)
W/python.stderr:        at android.os.Handler.dispatchMessage(Handler.java:101)
W/python.stderr:        at android.os.Looper.loop(Looper.java:166)
W/python.stderr:        at android.app.ActivityThread.main(ActivityThread.java:7529)
W/python.stderr:        at java.lang.reflect.Method.invoke(Native Method)
W/python.stderr:        at com.android.internal.os.Zygote$MethodAndArgsCaller.run(Zygote.java:245)
W/python.stderr:        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:921)

**Reply by freakboy3742** [reply]
This time, the problem is that your app doesn't have the necessary permissions declared. Your `pyproject.toml` needs to declare that it is going to use the Camera service (see the Hardware example project or the documentation for the Camera API for details). 

If you've added permissions to a project that you've previously generated, you'll need to delete the `build` folder and re-run `briefcase create android` to get the new permissions declared, or manually update the AndroidManifest.xml that already exists.

**Reply by jlij** [reply]
```
<?xml version="1.0" encoding="utf-8"?>
<manifest xmlns:android="http://schemas.android.com/apk/res/android" >
    <uses-permission android:name="android.permission.INTERNET" />
    <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />
    <uses-permission android:name="android.permission.CAMERA" />
    <uses-feature android:name="android.hardware.camera" android:required="false" />
    <uses-feature android:name="android.hardware.camera.any" android:required="false" />
    <uses-feature android:name="android.hardware.camera.front" android:required="false" />
    <uses-feature android:name="android.hardware.camera.external" android:required="false" />
    <uses-feature android:name="android.hardware.camera.autofocus" android:required="false" />
    <application
        android:allowBackup="true"
        android:icon="@mipmap/ic_launcher"
        android:label="@string/formal_name"
        android:networkSecurityConfig="@xml/network_security_config"
        android:roundIcon="@mipmap/ic_launcher_round"
        android:supportsRtl="true"
        android:theme="@style/AppTheme.Launcher" >
        <!-- https://developer.android.com/guide/topics/resources/runtime-changes#HandlingTheChange -->
        <activity
            android:configChanges="orientation|screenSize|screenLayout|keyboardHidden"
            android:name="org.beeware.android.MainActivity"
            android:exported="true" >
            <intent-filter>
                <action android:name="android.intent.action.MAIN" />
                <category android:name="android.intent.category.LAUNCHER" />
            </intent-filter>
            
        </activity>
        <provider
            android:name="androidx.core.content.FileProvider"
            android:authorities="com.example.helloworld.fileprovider"
            android:exported="false"
            android:grantUriPermissions="true">
            <meta-data
                android:name="android.support.FILE_PROVIDER_PATHS"
                android:resource="@xml/file_paths">
            </meta-data>
        </provider>
        
    </application>
    
</manifest>

```

**Reply by jlij** [reply]
pyproject.toml 

`# This project was generated with 0.3.17 using template: https://github.com/beeware/briefcase-template@v0.3.17
[tool.briefcase]
project_name = "Hello World"
bundle = "com.example"
version = "0.0.1"
url = "https://example.com/helloworld"
license = "BSD license"
author = "Jane Developer"
author_email = "jane@example.com"

[tool.briefcase.app.helloworld]
formal_name = "Hello World"
description = "My first application"
long_description = """More details about the app should go here.
"""
icon = "src/helloworld/resources/helloworld"
sources = [
    'src/helloworld',
]
test_sources = [

]
permission.camera = 'android.permission.CAMERA'
requires = [
]

test_requires = ["pytest", ]


[tool.briefcase.app.helloworld.macOS]
universal_build = true
requires = [
    "toga-cocoa~=0.4.0",
    "std-nslog~=1.0.0",
]


[tool.briefcase.app.helloworld.linux]
requires = [
    "toga-gtk~=0.4.0",
]


[tool.briefcase.app.helloworld.linux.system.debian]
system_requires = [
    # Needed to compile pycairo wheel
    "libcairo2-dev",
    # Needed to compile PyGObject wheel
    "libgirepository1.0-dev",
]

system_runtime_requires = [
    # Needed to provide GTK and its GI bindings
    "gir1.2-gtk-3.0",
    "libgirepository-1.0-1",
    # Dependencies that GTK looks for at runtime
    "libcanberra-gtk3-module",
    # Needed to provide WebKit2 at runtime
    # "gir1.2-webkit2-4.0",
]


[tool.briefcase.app.helloworld.linux.system.rhel]
system_requires = [
    # Needed to compile pycairo wheel
    "cairo-gobject-devel",
    # Needed to compile PyGObject wheel
    "gobject-introspection-devel",
]

system_runtime_requires = [
    # Needed to support Python bindings to GTK
    "gobject-introspection",
    # Needed to provide GTK
    "gtk3",
    # Dependencies that GTK looks for at runtime
    "libcanberra-gtk3",
    # Needed to provide WebKit2 at runtime
    # "webkit2gtk3",
]


[tool.briefcase.app.helloworld.linux.system.suse]
system_requires = [
    # Needed to compile pycairo wheel
    "cairo-devel",
    # Needed to compile PyGObject wheel
    "gobject-introspection-devel",
]

system_runtime_requires = [
    # Needed to provide GTK
    "gtk3",
    # Needed to support Python bindings to GTK
    "gobject-introspection", "typelib(Gtk) = 3.0",
    # Dependencies that GTK looks for at runtime
    "libcanberra-gtk3-0",
    # Needed to provide WebKit2 at runtime
    # "libwebkit2gtk3",
    # "typelib(WebKit2)",
]


[tool.briefcase.app.helloworld.linux.system.arch]
system_requires = [
    # Needed to compile pycairo wheel
    "cairo",
    # Needed to compile PyGObject wheel
    "gobject-introspection",
    # Runtime dependencies that need to exist so that the
    # Arch package passes final validation.
    # Needed to provide GTK
    "gtk3",
    # Dependencies that GTK looks for at runtime
    "libcanberra",
    # Needed to provide WebKit2
    # "webkit2gtk",
]

system_runtime_requires = [
    # Needed to provide GTK
    "gtk3",
    # Needed to provide PyGObject bindings
    "gobject-introspection-runtime",
    # Dependencies that GTK looks for at runtime
    "libcanberra",
    # Needed to provide WebKit2 at runtime
    # "webkit2gtk",
]


[tool.briefcase.app.helloworld.linux.appimage]
manylinux = "manylinux_2_28"

system_requires = [
    # Needed to compile pycairo wheel
    "cairo-gobject-devel",
    # Needed to compile PyGObject wheel
    "gobject-introspection-devel",
    # Needed to provide GTK
    "gtk3-devel",
    # Dependencies that GTK looks for at runtime, that need to be
    # in the build environment to be picked up by linuxdeploy
    "libcanberra-gtk3",
    "PackageKit-gtk3-module",
    "gvfs-client",
]

linuxdeploy_plugins = [
    "DEPLOY_GTK_VERSION=3 gtk",
]


[tool.briefcase.app.helloworld.linux.flatpak]
flatpak_runtime = "org.gnome.Platform"
flatpak_runtime_version = "45"
flatpak_sdk = "org.gnome.Sdk"


[tool.briefcase.app.helloworld.windows]
requires = [
    "toga-winforms~=0.4.0",
]


# Mobile deployments
[tool.briefcase.app.helloworld.iOS]
requires = [
    "toga-iOS~=0.4.0",
    "std-nslog~=1.0.0",
]


[tool.briefcase.app.helloworld.android]
requires = [
    "toga-android==0.4.2"
]

base_theme = "Theme.MaterialComponents.Light.DarkActionBar"

build_gradle_dependencies = [
    "androidx.appcompat:appcompat:1.6.1",
    "com.google.android.material:material:1.11.0",
    # Needed for DetailedList
    "androidx.swiperefreshlayout:swiperefreshlayout:1.1.0",
]


# Web deployments
[tool.briefcase.app.helloworld.web]
requires = [
    "toga-web~=0.4.0",
]
style_framework = "Shoelace v2.3"


`

**Reply by jlij** [reply]
Could you please help me check where the problem is?

**Reply by freakboy3742** [reply]
I can't see anything obviously wrong there - as far as I can tell, it's the default Briefcase output for a HelloWorld app that declares a photo permission. If you're getting an error that says it can't find a FileProvider, it suggests that the version of the app you're running hasn't been recompiled - it's using the old version. To be completely sure, manually delete the app from your phone, then run `briefcase run android` to re-install it.

**Reply by jlij** [reply]
You are right, it did appear.
W/python.stderr: java.lang.IllegalArgumentException: Couldn't find meta-data for provider with authority org.beeware.examples.hardware.fileprovider

I have recompiled according to your method, but it did not work。

**Reply by jlij** [reply]
briefcase run android  Subsequent prompts


E/vndksupport: Could not load vendor/lib64/egl/libGLES_mali.so from sphal namespace: dlopen failed: library "vendor/lib64/egl/libGLES_mali.so" not found.



I/PressGestureDetector: onAttached end
I/PressGestureDetector: enabledInPad = false,isPcCastMode = false
I/Adreno  : QUALCOMM build                   : cf751ce, Ia11ce2d146
I/Adreno  : Build Date                       : 12/22/20
I/Adreno  : OpenGL ES Shader Compiler Version: EV031.20.00.04
I/Adreno  : Local Branch                     :
I/Adreno  : Remote Branch                    : refs/tags/AU_LINUX_ANDROID_LA.UM.6.5.R1.08.00.00.312.030
I/Adreno  : Remote Branch                    : NONE
I/Adreno  : Reconstruct Branch               : NOTHING
I/Adreno  : PFP: 0x005ff087, ME: 0x005ff063
I/OpenGLRenderer: Initialized EGL, version 1.4
D/OpenGLRenderer: Swap behavior 2
E/vndksupport: Could not load vendor/lib64/egl/libGLES_mali.so from sphal namespace: dlopen failed: library "vendor/lib64/egl/libGLES_mali.so" not found.
D/OpenGLRenderer:   HWUI Binary is  disabled
D/MainActivity: onPrepareOptionsMenu() start
D/MainActivity: onPrepareOptionsMenu() complete
W/InputMethodManager: startInputReason = 1
W/InputMethodManager: startInputReason = 5
I/HwApiCacheMangerEx: apicache pi PackageInfo{6fb42a2 com.example.helloworld}
D/ProfileInstaller: Installing profile for com.example.helloworld
read: unexpected EOF!

**Reply by freakboy3742** [reply, marked as answer]
If you're a Briefcase generated app template that generated the configuration files you provided earlier, you *can't* be getting that error. The error message talks about `org.beeware.examples.hardware.fileprovider`; your application is `com.example.helloworld`. 

I strongly suspect you've copy-and-pasted some code from the Hardware example - most likely the code that sets the app's bundle ID. If you correct the bundle ID and rebuild, it should find the correct file provider.

**Reply by jlij** [reply]
Thank you very much. After deleting the entire project and rebuilding it, we will solve it perfectly according to the instructions you provided.

**Comment by yyz-git** [comment]
How can this error be resolved?

W/python.stderr: Error in async handler: WARNING ⚠️ Insufficient free disk space 0.0 GB < 0.000 GB required, Please free 0.0 GB additional disk space and try again.
W/python.stderr: Traceback (most recent call last):
W/python.stderr:   File "/data/data/com.example.pose_app/files/chaquopy/AssetFinder/requirements/toga/handlers.py", line 95, in handler_with_cleanup
W/python.stderr:   File "/data/data/com.example.pose_app/files/chaquopy/AssetFinder/app/pose_app/app.py", line 37, in take_photo
W/python.stderr:   File "/data/data/com.example.pose_app/files/chaquopy/AssetFinder/requirements/ultralytics/models/yolo/model.py", line 23, in __init__
W/python.stderr:   File "/data/data/com.example.pose_app/files/chaquopy/AssetFinder/requirements/ultralytics/engine/model.py", line 145, in __init__
W/python.stderr:   File "/data/data/com.example.pose_app/files/chaquopy/AssetFinder/requirements/ultralytics/engine/model.py", line 285, in _load
W/python.stderr:   File "/data/data/com.example.pose_app/files/chaquopy/AssetFinder/requirements/ultralytics/nn/tasks.py", line 910, in attempt_load_one_weight
W/python.stderr:   File "/data/data/com.example.pose_app/files/chaquopy/AssetFinder/requirements/ultralytics/nn/tasks.py", line 815, in torch_safe_load
W/python.stderr:   File "/data/data/com.example.pose_app/files/chaquopy/AssetFinder/requirements/ultralytics/utils/downloads.py", line 454, in attempt_download_asset
W/python.stderr:   File "/data/data/com.example.pose_app/files/chaquopy/AssetFinder/requirements/ultralytics/utils/downloads.py", line 331, in safe_download
W/python.stderr:   File "/data/data/com.example.pose_app/files/chaquopy/AssetFinder/requirements/ultralytics/utils/downloads.py", line 231, in check_disk_space
W/python.stderr: MemoryError: WARNING ⚠️ Insufficient free disk space 0.0 GB < 0.000 GB required, Please free 0.0 GB additional disk space and try again.

**Reply by freakboy3742** [reply]
Well... I'd suggest doing what the error message suggests, and freeing some disk space - that's the space on the *device*, not your development machine.

Beyond that, to be able to help you we'd need to know *anything* about what you've done to generate that error. 

The fact that the stack trace references "torch" suggests you're trying to load an AI model into your app - it's very likely that you're exceeding the 512MB size limit for Android app bundles.
